### ISSUE TITLE
[dataio] I3FrameSequence handling of truncated files (Trac #2106)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#2098]
I frequently find it useful to be able to view the portion of an I3 file which was written up to when the writing process crashed. However, this often involves the file containing a damaged final frame, which was incompletely written. My recollection is that `dataio-shovel` used to handle this fine (although I could be wrong), but now that it is based on `I3FrameSequence` it does not. Ideally, I would like to have the option to treat a malformed frame, beyond which reading is not possible, as simply representing the effective end of the file, and possibly logging an error. The following patch is sufficient to make this work for `dataio-shovel`, but am not certain that it treats all cases and leaves all datastructures in a fully consistent state:

```text
Index: private/dataio/I3FrameSequence.cxx
===================================================================
--- private/dataio/I3FrameSequence.cxx	(revision 3000)
+++ private/dataio/I3FrameSequence.cxx	(working copy)
@@ -487,13 +487,21 @@
             // check future cache
             const auto tr_cache_iter = tr_cache_.find(frameno_);
             if (tr_cache_iter != tr_cache_.end()) {
-                fr = tr_cache_iter->second.get(); // get from future
+                try {
+                    fr = tr_cache_iter->second.get(); // get from future
+                } catch(icecube::archive::archive_exception& ex) {
+                    log_warn_stream("Unexpected end of input file: " << ex.what());
+                }
                 tr_cache_.erase(tr_cache_iter);
             } else {
                 // make future, then wait on it
-                auto f = tr_.push(
-                    std::bind(&FileGroup::get_frame, std::ref(files_), frameno_));
-                fr = f.get();
+                auto f = tr_.push([=]{return files_.get_frame(frameno_);});
+                try {
+                    fr = f.get();
+                } catch(icecube::archive::archive_exception& ex) {
+                    log_warn_stream("Unexpected end of input file: " << ex.what());
+                    return false;
+                }
             }
             if (!fr.empty()) {
                 // add to frame cache
```

Migrated from https://code.icecube.wisc.edu/ticket/2106
```json
{
    "status": "closed", 
    "changetime": "2019-02-13T14:15:23", 
    "description": "I frequently find it useful to be able to view the portion of an I3 file which was written up to when the writing process crashed. However, this often involves the file containing a damaged final frame, which was incompletely written. My recollection is that `dataio-shovel` used to handle this fine (although I could be wrong), but now that it is based on `I3FrameSequence` it does not. Ideally, I would like to have the option to treat a malformed frame, beyond which reading is not possible, as simply representing the effective end of the file, and possibly logging an error. The following patch is sufficient to make this work for `dataio-shovel`, but am not certain that it treats all cases and leaves all datastructures in a fully consistent state:\n\n{{{\nIndex: private/dataio/I3FrameSequence.cxx\n===================================================================\n--- private/dataio/I3FrameSequence.cxx\t(revision 3000)\n+++ private/dataio/I3FrameSequence.cxx\t(working copy)\n@@ -487,13 +487,21 @@\n             // check future cache\n             const auto tr_cache_iter = tr_cache_.find(frameno_);\n             if (tr_cache_iter != tr_cache_.end()) {\n-                fr = tr_cache_iter->second.get(); // get from future\n+                try {\n+                    fr = tr_cache_iter->second.get(); // get from future\n+                } catch(icecube::archive::archive_exception& ex) {\n+                    log_warn_stream(\"Unexpected end of input file: \" << ex.what());\n+                }\n                 tr_cache_.erase(tr_cache_iter);\n             } else {\n                 // make future, then wait on it\n-                auto f = tr_.push(\n-                    std::bind(&FileGroup::get_frame, std::ref(files_), frameno_));\n-                fr = f.get();\n+                auto f = tr_.push([=]{return files_.get_frame(frameno_);});\n+                try {\n+                    fr = f.get();\n+                } catch(icecube::archive::archive_exception& ex) {\n+                    log_warn_stream(\"Unexpected end of input file: \" << ex.what());\n+                    return false;\n+                }\n             }\n             if (!fr.empty()) {\n                 // add to frame cache\n}}}\n", 
    "reporter": "cweaver", 
    "cc": "", 
    "resolution": "fixed", 
    "_ts": "1550067323910946", 
    "component": "combo core", 
    "summary": "[dataio] I3FrameSequence handling of truncated files", 
    "priority": "minor", 
    "keywords": "", 
    "time": "2017-10-25T21:36:03", 
    "milestone": "", 
    "owner": "olivas", 
    "type": "enhancement"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20171025T21:47:23: david.schultz commented:

> I think that's mostly fine, except the first catch also needs to return false and remove the tr_cache_iter from tr_cache_.

**Comment by icecube-trac** [comment]
Trac update at 20171025T21:51:41: cweaver commented:

> That should happen due to the existing control flow after the catch, no? I.e. the `erase` will excute next, and that block was only entered if `fr.empty()`, so `false` will be returned at the end of the function.

**Comment by icecube-trac** [comment]
Trac update at 20171025T21:55:36: david.schultz commented:

> Hmm, I guess so.  It takes a few more comparisons, but works fine.

**Comment by icecube-trac** [comment]
Trac update at 20180922T09:54:12: olivas changed milestone from "" to "Winter Solstice 2018"

**Comment by icecube-trac** [comment]
Trac update at 20180922T12:20:09: 
* olivas changed owner from "david.schultz" to "olivas"
* olivas changed status from "new" to "assigned"

**Comment by icecube-trac** [comment]
Trac update at 20180927T08:29:50: david.schultz commented:

> One of the things this might do is hide errors in actual user code.  Not sure how to normally raise an exception, except for the shovel.

**Comment by icecube-trac** [comment]
Trac update at 20181204T01:16:24: olivas changed status from "assigned" to "accepted"

**Comment by icecube-trac** [comment]
Trac update at 20181205T02:28:40: 
* olivas changed resolution from "" to "fixed"
* olivas changed status from "accepted" to "closed"
* olivas commented:

> In changeset 3113:
> 
> ```text
> #!CommitTicketReference repository="IceTray" revision="3113"
> all dataio tests pass.  let's see if we can recover from corrupt files.  closes #2098.
> ```

**Comment by icecube-trac** [comment]
Trac update at 20190111T22:02:52: nwhitehorn commented:

> In changeset 169268:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="169268"
> all dataio tests pass.  let's see if we can recover from corrupt files.  closes #2098.
> ```

**Comment by icecube-trac** [comment]
Trac update at 20190213T14:15:23: 
* olivas commented:

> Milestone Winter Solstice 2018 deleted


* olivas changed milestone from "Winter Solstice 2018" to ""
